opentreemap/treemap/instance.py

```diff
@@ -768,9 +771,39 @@ def raise_errors(errors):
         if errors:
             raise_errors(errors)
 
-    def save(self, *args, **kwargs):
+    @transaction.atomic
+    def _bulk_assign_task_reference_numbers(self, tasks):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for task in tasks:
+            task.reference_number = instance.task_sequence_number
+            instance.task_sequence_number += 1
+        instance.save(update_fields=['task_sequence_number'])
+
+    @transaction.atomic
+    def _bulk_assign_work_order_reference_numbers(self, work_orders):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for work_order in work_orders:
+            work_order.reference_number = instance.work_order_sequence_number
+            instance.work_order_sequence_number += 1
+        instance.save(update_fields=['work_order_sequence_number'])
+
+    def save(self, tasks=None, work_orders=None, *args, **kwargs):
+        """
+        Keyword Arguments:
+
+        tasks - List of Task models
+        work_orders - List of WorkOrder models
+        """
         self.full_clean()
 
         self.url_name = self.url_name.lower()
 
+        if tasks:
```

Comment:
This `Instance` helper looks like it may be a good starting point

```python
@transaction.atomic
def get_next_task_sequence(self, n=1):
    if n < 1:
        raise ValueError('n must be >= 1')
    qs = Instance.objects.select_for_update()
    next_number = qs.get(id=self.id).work_order_sequence_number
    qs.filter(id=self.id).update(
        work_order_sequence_number=F('work_order_sequence_number') + n)
    if 1 == n:
        return next_number
    else:
        return range(next_number, next_number + n)
```

```
In [1]: i1 = Instance.objects.get(url_name='phillytreemap')

In [2]: i2 = Instance.objects.get(url_name='phillytreemap')

In [3]: i1.get_next_task_sequence()
Out[3]: 1

In [4]: i2.get_next_task_sequence()
Out[4]: 2

In [5]: i1.get_next_task_sequence(n=10)
Out[5]: [3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

In [6]: i2.get_next_task_sequence()
Out[6]: 13
```